manta-accounting/src/wallet/signer/functions.rs

```diff
@@ -774,6 +808,137 @@ where
+/// returning [`TransferPost`]s.
+#[allow(clippy::too_many_arguments)]
+#[inline]
+fn compute_to_public_transaction<C>(
+    accounts: &AccountTable<C>,
+    assets: &C::AssetMap,
+    utxo_accumulator: &mut C::UtxoAccumulator,
+    parameters: &Parameters<C>,
+    proving_context: &MultiProvingContext<C>,
+    asset_id: &C::AssetId,
+    sink_accounts: Vec<C::AccountId>,
+    selection: Selection<C>,
+    rng: &mut C::Rng,
+) -> Result<SignResponse<C>, SignError<C>>
+where
+    C: Configuration,
+    C::AssetValue: SubAssign,
+{
+    let Selection {
+        mut change,
+        mut pre_senders,
+    } = selection;
+    let mut posts = Vec::new();
+    let mut iter = pre_senders
+        .into_iter()
+        .chunk_by::<{ ToPublicShape::SENDERS }>();
+    for chunk in &mut iter {
+        let senders = array_map(chunk, |s| {
+            s.try_upgrade(parameters, utxo_accumulator)
+                .expect("Unable to upgrade expected UTXO.")
+        });
+        process_to_public_senders(
+            accounts,
+            utxo_accumulator,
+            parameters,
+            proving_context,
+            asset_id,
+            &mut change,
+            senders,
+            sink_accounts.clone(),
+            &mut posts,
+            rng,
+        )?;
+    }
+    pre_senders = iter.remainder();
+    if !pre_senders.is_empty() {
+        let final_senders = into_array_unchecked(prepare_final_pre_senders(
+            accounts,
+            assets,
+            utxo_accumulator,
+            parameters,
+            asset_id,
+            Default::default(),
+            pre_senders,
+            rng,
+        )?);
+        process_to_public_senders(
+            accounts,
+            utxo_accumulator,
+            parameters,
+            proving_context,
+            asset_id,
+            &mut change,
+            final_senders,
+            sink_accounts,
+            &mut posts,
+            rng,
+        )?;
+    }
+    Ok(SignResponse::new(posts))
+}
+
+/// Creates a to public [`TransferPost`] spending the assets held by `senders` and
+/// attaches it to `post`.
+#[allow(clippy::too_many_arguments)]
+#[inline]
+fn process_to_public_senders<C>(
+    accounts: &AccountTable<C>,
+    utxo_accumulator: &mut C::UtxoAccumulator,
+    parameters: &Parameters<C>,
+    proving_context: &MultiProvingContext<C>,
+    asset_id: &C::AssetId,
+    change: &mut C::AssetValue,
+    senders: [Sender<C>; ToPublicShape::SENDERS],
+    sink_accounts: Vec<C::AccountId>,
+    posts: &mut Vec<TransferPost<C>>,
+    rng: &mut C::Rng,
+) -> Result<(), SignError<C>>
+where
+    C: Configuration,
+    C::AssetValue: SubAssign,
+{
+    let authorization = authorization_for_default_spending_key::<C>(accounts, parameters, rng);
+    let mut received_value = C::AssetValue::default();
+    let mut reclaimed_value = senders
+        .iter()
+        .map(|sender| sender.asset().value)
+        .sum::<C::AssetValue>();
+    if reclaimed_value >= *change {
```

Comment:
Ok this is the common case when eg: i have a utxo of 100 but want 90 , so have change 10.
Not getting the else path though when the reclaimed_value is supposed to end up 0 ?

Reply:
This function is called only by `compute_to_public_transaction`, which in turn is only called after a successful `selection` call, meaning the total amount of asset in the (pre)senders will always be at least `change` (the output of `selection`). Now this function is called for each pair of senders (and not for all of them at once). For each pair, if their total value is larger than `change`, we:
1) create a receiver with `change` amount of asset. 
2) set `change` to 0, causing all subsequent calls to this function will reclaim the whole amount
3) reclaim the difference

Note that as a result of `compute_to_public_transaction` being only called with a valid `selection`, we will hit this path at least once in the last call.

In the `else` case, we
1) don't reclaim anything 
2) create a receiver with all the value in the senders
3) subtract that value from `change` (because we've changed it already when we created the receiver).

I believe the logic is sound.